[Ng Geon Woo, Robin] iP

Oddcorner opened the pull request:
# Duke Task Manager
> "This application gets 10 apples out of 10!" - Apple farmer ([source](https://www.apple.com/))

Duke Task Manager is a simple to use text-based task manager that allows you to:
- Remember important events and deadlines
- Mark tasks as done
- Find tasks
- And is ~~EASY~~ **SUPER EASY** to use

Some commands include:
1. list
2. mark/unmark
4. find
5. delete

Just download the jar file and run it with the command: `java -jar Main.jar`

Features planned:
- [X] Managing tasks 👍 
- [ ] Reminding of upcoming deadlines 📆 
- [ ] Multi-user support 🧑‍🤝‍🧑 

---
Wanna get your hands _dirty_ by practising java? Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Oddcorner committed:
Update Duke.java

Oddcorner committed:
Add list feature

Oddcorner committed:
Add Task.java

Oddcorner committed:
Add "Mark as done" feature

Oddcorner committed:
Add Deadline.java, Event.java, Todo.java

Oddcorner committed:
Update Duke.java to handle new Tasks Types

Oddcorner committed:
Change line break design

Oddcorner committed:
Add ui-test files

Oddcorner committed:
Add exception handling of user inputs

Oddcorner committed:
Add delete function

Oddcorner committed:
Refactor Duke.java

Oddcorner committed:
Add "Save/Load" of list data

Currently the Arraylist of Tasks in Duke.java is lost after the program exits.

This makes the program vulnerable to power loss and accidental exits.

Therefore this commit updates Duke.java to save the Arraylist data to "data/data.bin" and load from it if the file exists on start.
It also updates Task.java to make the class implement Serailizable for this purpose.

Oddcorner committed:
Fix "directory not found"

Oddcorner committed:
Update Deadline.java to store dates as LocalDate

Currently deadline task stores dates as a String type.

This allows for undesirable inputs as format is not enforced.

This commit changes this String field to a LocalDate type, allowing for more functionality in the future

Oddcorner committed:
Merge branch 'branch-Level-7'

Oddcorner committed:
Merge branch 'branch-Level-8'

Oddcorner committed:
Add Command class and its subclasses

Currently all the functionality of the program is handled internally in
Duke.java

This  increases the difficultly in adding and editing commands while
maintaining code correctness.

Therefore, this commit will add a new Command class as well as sub
classes for the various commands to make the project more
Object oriented.

Oddcorner committed:
Add Parser.java

This is a new class that handles parsing String input from users and
returns the appropriate Command object.

Oddcorner committed:
Add Storage.java

This class contains the functions to handle the storage and retrieval of
the task list.

Oddcorner committed:
Add TaskList.java

Oddcorner committed:
Add Ui.java

Class that defines and prints messages to the screen.

Oddcorner committed:
Refactor Duke.java

Duke class handles all the logic in the program from parsing, storage,
printing, and managing an internal list of Tasks.

Adding new features will prove challenging as the project grows.

Let's separate the functionality of Duke into individual classes
following a more OOP design.

Oddcorner committed:
Move classes into suitable java packages

Oddcorner committed:
Add JUnit tests

Oddcorner committed:
Add JavaDoc comments

Oddcorner committed:
Tweak code to comply with coding standard

Oddcorner committed:
Add FindCommand.java

Task list can get arbitrarily long as users add more tasks.

This makes locating the index of a desired task challenging.

Let's,
* add a new command class called FindCommand that allows users
to search for tasks that is in the task list.
* update Ui.java to support new messages generated by this command.
* update Parser.java to parse user input "find" to the new command.

Oddcorner committed:
Merge branch 'branch-A-JavaDoc'

Oddcorner committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/tasklist/TaskList.java
#	src/main/java/duke/ui/Ui.java

Oddcorner committed:
Merge branch 'branch-Level-9'

Oddcorner committed:
Merge pull request #2 from Oddcorner/add-gradle-support

Add Gradle support

Oddcorner committed:
Add Gradle support

Oddcorner committed:
Merge branch 'branch-A-Gradle'

NatalieTanML commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.time.LocalDate;
+
+public class AddCommand extends Command {
```
I think you missed out on the Javadoc for this class

Fenway17 replied:
Short javadoc description can be added here.

NatalieTanML commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,127 @@
+     *
+     * @param tasks List of tasks to be printed.
+     */
+    public void list(List<Task> tasks) {
```
Would naming this `printList` align with the code guidelines better?

Oddcorner replied:
Youre right! 🦖

NatalieTanML commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,127 @@
+     * @param task Task that was added.
+     * @param noOfTasks Number of tasks that are currently in the list of tasks.
+     */
+    public void taskAddedMessage(Task task, int noOfTasks) {
```
Would `printAddedMessage` align with the code quality guidelines more?

NatalieTanML commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,127 @@
+     *
+     * @param task Task that was successfully marked.
+     */
+    public void taskMarkedMessage(Task task) {
```
Similarly for these other method names; they are under the Ui class but I think it would be better if they were more [descriptive](https://nus-cs2103-ay2122s2.github.io/website/se-book-adapted/chapters/codeQuality.html#use-name-to-explain)

NatalieTanML commented on .gitignore:
```diff
@@ -15,3 +15,15 @@ bin/
 
 /text-ui-test/ACTUAL.txt
 text-ui-test/EXPECTED-UNIX.TXT
+data/data.bin
+src/main/java/duke/.idea/.gitignore
+src/main/java/duke/.idea/misc.xml
+src/main/java/duke/.idea/modules.xml
+src/main/java/duke/.idea/vcs.xml
+src/main/java/duke/.idea/codeStyles/codeStyleConfig.xml
+src/main/java/duke/data/data.bin
+text-ui-test/data/data.bin
+src/test/.idea/.gitignore
+src/test/.idea/misc.xml
+src/test/.idea/modules.xml
+src/test/.idea/vcs.xml
```
This is a minor thing but you can shorten these by just including the directory, e.g. `data/`, `src/test/.idea/` and the files under the directory will be ignored automatically

NatalieTanML reviewed (commented):
Overall the code is quite clean and easy to understand, just that its missing Javadoc for quite a few of the classes. LGTM!

bakano98 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,19 @@
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
```
You may have missed out JavaDoc for this class

bakano98 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,68 @@
+ */
 public class Duke {
+    public boolean isListening = true;
+    private final String FILE_PATH;
```
Good to see you're following the naming convention for constants! Would setting this to a pre-defined, hardcoded value be better than defining in your `main` method?

bakano98 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,79 @@
+                FILE.createNewFile();
+            }
+        } catch (IOException e) {
+            System.out.println("Something went wrong when creating initialising the file: "
```
Good to see you breaking lines when it's too long. Keep it up! 😄

bakano98 commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,12 @@
+package duke.task;
+
+public class Todo extends Task {
```
Should this be named `ToDo` since it's two words?

bakano98 reviewed (commented):
Very good overall. Follows the [coding standards](https://se-education.org/guides/conventions/java/intermediate.html) well. LGTM!

Oddcorner committed:
Add GUI for Duke

Duke only supports a text based UI.

This makes the program unappealing and dull to use as it is hard
to convey character through text alone.

A GUI helps make the program more appealing to use and allows for
creative design choices in the future.

Let's do the following to add a GUI for Duke using JavaFX,
* Modify the Command classes to allow them to modify StringBuilder
objects.
* Move Duke.java into the duke package.
* Change the return type of Ui methods from void to String to pass
to the new GUI classes to display.
* Modify UiTest.java to reflect the changes made in Ui.java.
* Add new GUI classes: Main, Launcher, DialogBox, MainWindow.

Oddcorner committed:
Merge branch 'branch-Level-10'

Fenway17 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,51 @@
+    private String description;
+    private LocalDate date;
+
+    public AddCommand(String keyword, String description, LocalDate date) {
+        super(keyword);
+        this.description = description;
+        this.date = date;
+    }
+
+    public AddCommand(String keyword, String description) {
+        super(keyword);
+        this.description = description;
+        this.date = null;
+    }
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage, StringBuilder response) {
+        Task task = null;
+        switch (keyword) {
+        case "todo":
+            task = new Todo(description);
+            break;
+        case "event":
+            task = new Event(description, date);
+            break;
+        case "deadline":
+            task = new Deadline(description, date);
+            break;
+        default:
+            break;
+        }
+        if (task != null) {
+            tasks.add(task);
+            storage.save(tasks.list());
+            response.append(ui.taskAddedMessage(task, tasks.size()));
+        }
+    }
```
Perhaps javadocs can be added for these constructors and methods.

Fenway17 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,84 @@
+        String[] strArr = input.split(" ", 2);
+        String keyword = strArr[0];
+
+        Command command = new Command("default");
```
I like the use of switch statements here, easy to understand!

Fenway17 reviewed (commented):
Code is overall clean and not confusing. Only nits are missing javadocs for some classes and methods.

Oddcorner committed:
Add java assertions

Oddcorner committed:
Refactor Command classes

Several command classes do not follow code quality guidelines
hindering the readability of the code.

By refactoring these command classes following some good code
quality principles, we can improve the readability and
maintainability of the code.

Let's,
* Extract the method for checking out of bounds index in:
  * DeleteCommand.java
  * MarkCommand.java
  * UnmarkCommand.java
* Extract the method for getting search results in FindCommand.java
to follow SLAP

Oddcorner committed:
Refactor DialogBox.java

Oddcorner committed:
Merge pull request #3 from Oddcorner/branch-A-Assertions

Add java assertions

Oddcorner committed:
Merge branch 'master' of https://github.com/Oddcorner/ip

Oddcorner committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/UnmarkCommand.java

Oddcorner committed:
Merge pull request #4 from Oddcorner/branch-A-CodeQuality

Branch a code quality

Oddcorner committed:
Merge branch 'master' of https://github.com/Oddcorner/ip

Oddcorner committed:
Add help command

Oddcorner committed:
Merge branch 'branch-Extensions'

Oddcorner committed:
Refactor: Change name from Duke to Taskie

Oddcorner committed:
Update README.md

Oddcorner committed:
Set theme jekyll-theme-architect

Oddcorner committed:
Add more javadocs

Oddcorner committed:
Merge branch 'master' of https://github.com/Oddcorner/ip
